x-pack/plugins/apm/server/routes/span_links/get_outgoing_span_links.ts

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import { rangeQuery } from '@kbn/observability-plugin/server';
+import moment from 'moment';
+import {
+  PROCESSOR_EVENT,
+  SPAN_ID,
+  SPAN_LINKS,
+  SPAN_LINKS_TRACE_ID,
+  SPAN_LINKS_SPAN_ID,
+  TRACE_ID,
+  TRANSACTION_ID,
+} from '../../../common/elasticsearch_fieldnames';
+import { ProcessorEvent } from '../../../common/processor_event';
+import type { SpanRaw } from '../../../typings/es_schemas/raw/span_raw';
+import type { TransactionRaw } from '../../../typings/es_schemas/raw/transaction_raw';
+import { Setup } from '../../lib/helpers/setup_request';
+
+async function fetchOutgoingSpans({
+  traceId,
+  setup,
+  start,
+  end,
+  spanId,
+}: {
+  traceId: string;
+  setup: Setup;
+  start: number;
+  end: number;
+  spanId?: string;
+}) {
+  const { apmEventClient } = setup;
+
+  const startWithBuffer = moment(start).subtract(4, 'days').valueOf();
+  const endWithBuffer = moment(end).add(4, 'days').valueOf();
+
+  const response = await apmEventClient.search('get_outgoing_span_links', {
+    apm: {
+      events: [ProcessorEvent.span, ProcessorEvent.transaction],
+    },
+    _source: [SPAN_LINKS, TRACE_ID, SPAN_ID, PROCESSOR_EVENT, TRANSACTION_ID],
+    body: {
+      size: 1000,
+      query: {
+        bool: {
+          filter: [
+            ...rangeQuery(startWithBuffer, endWithBuffer),
+            { term: { [SPAN_LINKS_TRACE_ID]: traceId } },
+            ...(spanId ? [{ term: { [SPAN_LINKS_SPAN_ID]: spanId } }] : []),
+          ],
+        },
+      },
+    },
+  });
+
+  return response.hits.hits;
+}
+
+function getSpanId(source: TransactionRaw | SpanRaw) {
+  return source.processor.event === ProcessorEvent.span
+    ? (source as SpanRaw).span.id
+    : (source as TransactionRaw).transaction?.id;
+}
+
+export async function getOutgoingSpanLinksCountBySpanId({
+  traceId,
+  setup,
+  start,
+  end,
+}: {
+  traceId: string;
+  setup: Setup;
+  start: number;
+  end: number;
+}) {
+  const outgoingSpans = await fetchOutgoingSpans({
+    traceId,
+    setup,
+    start,
+    end,
+  });
+
+  return outgoingSpans.reduce<Record<string, number>>(
+    (acc, { _source: source }) => {
+      source.span?.links?.forEach((link) => {
```

Comment:
that makes sense.

Reply:
@dgieselaar I gave it another thought and the current implementation is correct. 

That is because when I call `getOutgoingSpanLinksCountBySpanId` the `Transaction or Span` returned might have more links than the `traceId` used to filter the query, so I need to loop through the span links and only add to the Count map the links that belong to the current trace.

Now, this filter is not necessary for the `getOutgoingSpanLinks` because I'm not interested in the `span.links` from the transactions or spans returned, I'm interested in the `trace.id` and `transaction.id or span.id`.

Reply:
I don't follow. The purpose of this function is to return a list of linked child spans for the currently displayed transaction/span in the APM UI, correct? So you've got trace.id and span.id to query on. However, Elasticsearch will return you any **document** that has any span link with **either** the given trace.id or span.id, as long as they are both found at least once. Consider the following example:

We are looking for all linked child spans of span.id: span-A in trace.id: trace-A. We search for documents matching: `span.links.span.id:span-A AND span.links.trace.id:trace-A`. What we will get back is one with document with the following span links without filtering:

- ✅ `{ 'span.links.span.id': 'span-A', 'span.links.span.id': 'trace-A' }`
- ❌ `{ 'span.links.span.id': 'span-B', 'span.links.span.id': 'trace-A' }`
- ❌ `{ 'span.links.span.id': 'span-A', 'span.links.span.id': 'trace-B' }`
- ❌ `{ 'span.links.span.id': 'span-C', 'span.links.span.id': 'trace-C' }`

Maybe I'm missing something obvious but I can't quite grasp how this works without filtering the span links.